### PR TITLE
[COTCO-316] Integration testing + fixed aggregation queries to include/exclude expired records

### PR DESCRIPTION
**Pull request by robertbutacu**

### CODE REVIEW

File: test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala [hunk 1]
```diff
@@ -42,15 +45,45 @@ class TransitRepositorySpec extends MongoSetup with TransitGenerator with MongoS
...
       }
     }
+
+    "does not include any expired available staff records for enRoute transits" in new TransitRepositorySpecWithCleanMongo {
+      withActiveTransitRecord {
+        transit =>
+          val noTimeout = 0 // so I dont have to wait for 10 minutes before record becomes expired
+
+          await(transitRepository.create(Transit(transit)))
+          await(Future{Thread.sleep(1000)}) // used to make sure it makes the record expired
```
Comment (peppe-hmrc) [inline test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: Do we really need to have these flags and also wait to make the record expired?
I thought that given you have access to the repo you could just insert an expired record

Comment (peppe-hmrc) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: Anyway, I'd imagine it is ok too

File: it/uk/gov/hmrc/ctc/TransitControllerISpec.scala [hunk 2]
```diff
@@ -52,6 +52,34 @@ class TransitControllerISpec extends BaseISpec {
...
+    "return an empty list with 0 available staff for an empty database" in {
+      withNoSetup {
+        val result = callRoute(fakeRequest(routes.TransitController.aggregateData()))
+
+        status(result) shouldBe OK
+        val aggregation: TransitsAggregation = contentAsJson(result).as[TransitsAggregation]
+
+        // all the records are 0
+        assert(aggregation.enRouteTransits.forall(p => p.staffAvailable == 0 && p.estimations.forall(e => e.count == 0)))
+        assert(aggregation.arrivedTransits.forall(p => p.staffAvailable == 0 && p.estimations.forall(e => e.count == 0)))
```
Comment (peppe-hmrc) [inline it/uk/gov/hmrc/ctc/TransitControllerISpec.scala]: This is one of the reason why me and Greg were discussing to remove the priority gap filling in the BE.
In case of Empty dataset we send lists filled with `0`. I'm not sure an API should tell what is not present or what is present.
The idea was to move the filling in the FE (server side) however the API should include the number of priority slots used to made the calculations.
I thinks this would simplify the BE and also reduce the amount of data to send in the request for an empty dataset. What do you think about this approach?

File: app/uk/gov/hmrc/ctc/handlers/ErrorHandler.scala [hunk 3]
```diff
@@ -28,8 +28,8 @@ trait ErrorHandler {
   def toResult(error: CTCError): Result = {
     error match {
       case e: MalformedRequest => BadRequest(Json.toJson(ErrorResponse(400, e.message)))
-      case _: MongodbNoResults => NotFound(Json.toJson(ErrorResponse(404, "No results found for the request")))
-      case _: MongodbError => InternalServerError(Json.toJson(ErrorResponse(500, s"The request can't be fulfilled at the moment")))
+      case _: MongoDBNoResults => NotFound(Json.toJson(ErrorResponse(404, "No results found for the request")))
+      case _: MongoDBError => InternalServerError(Json.toJson(ErrorResponse(500, s"The request can't be fulfilled at the moment")))
```
Comment (peppe-hmrc) [inline app/uk/gov/hmrc/ctc/handlers/ErrorHandler.scala]: Just a small and personal comment about this renaming. 
I really appreciate it and you are right `MongoDBError` is much better and consistent name than `MongodbError`.
However, do you think that this was the right PR where to do it? I mean this PR has very important changes which touch the core business functionality. I have the impression that most of the file changes are only renaming.
Personally I would have made a separate PR for this. 
No hard feeling, this is just a personal comment, so feel free to ignore it :)

File: it/uk/gov/hmrc/ctc/TransitControllerISpec.scala (hunk 2 above)
Comment (robertbutacu) [reply it/uk/gov/hmrc/ctc/TransitControllerISpec.scala]: In either approach, we say what is present and what is not - in my case, it's explicit, in your case it's implicit.

To be honest, it sounds like moving complexity from one place to the other. 
It simplifies the BE, but it complicates the FE - and since its an aggregation, I'd imagine most use cases would want to treat the case when an aggregation is "missing" ( seen as 0 ) - in the current state, this is done ofc.

As far as the amount of data, we are talking about quantities incredibly small imo ( < 1 kbyte ?)

I'm open to discussions about this, but I'd rather have this as a tech debt and change when we have more free time.

File: app/uk/gov/hmrc/ctc/models/data/Transit.scala [hunk 4]
```diff
@@ -48,7 +48,8 @@ object Transit {
   implicit val format: OFormat[Transit] = OFormat(defaultReads,defaultWrites)
 
   val epochTimeBasedInstanceReads: Reads[Transit] = {
-    implicit val instantFormat: Format[ProcessingInfo] = ProcessingInfo.epochTimeFormat
+    implicit val processingInfoFormat: Format[ProcessingInfo] = ProcessingInfo.epochTimeFormat
+    implicit val recordFormat: Format[TransitRecord] = TransitRecord.epochBasedFormat
```
Comment (peppe-hmrc) [inline app/uk/gov/hmrc/ctc/models/data/Transit.scala]: Nice work, didn't know was that simple. 
Not sure if the default reader are still here, but I was wondering if we still need them given we decided to use `epochTimeFormat` everywhere. 
I was looking at `master` on the train and noticed the companion object is getting very bit :)

File: test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala (hunk 1 above)
Comment (robertbutacu) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: Should find a better way for this - Im not a fan either. 

Will think about this - we have access to mongo collection, but its a bit more complex than this

I ran the tests a couple of times without the Thread.sleep - which was just a cautionary measure to make sure they expire - and they seem to pass.

File: app/uk/gov/hmrc/ctc/models/data/Transit.scala (hunk 4 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/models/data/Transit.scala]: Thanks

Unfortunately, yes - the epochTimeFormat is used only for the repository level, while at an API level, it's still an Instant. 

We're stuck with the _id vs id fields, and thats why we can't have the default one.

Im up for changing the naming and be based on the simple Json.format[Transit] one - this implies changing it on the frontend as well.

File: test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala (hunk 1 above)
Comment (peppe-hmrc) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: I don't remember how I was inserting expired records ... It might worth look a it ... 
I think the problem you have is because you want to randomise it and about this approach I would say that not alway we want to randomise. Sometimes is a matter or a very specific case or edge case we want to test. So I'm not sure in these cases random is required

File: it/uk/gov/hmrc/ctc/TransitControllerISpec.scala (hunk 2 above)
Comment (peppe-hmrc) [reply it/uk/gov/hmrc/ctc/TransitControllerISpec.scala]: well it will be tech debt once we build the converter in the FE. Given this need to be implemented I would look better into it. I personally don't see the point of sending a list full of `0`. We (BE) are assuming how the FE would display it. 
If nobody else will raise a concern I'm happy the way it is

File: app/uk/gov/hmrc/ctc/models/data/Transit.scala (hunk 4 above)
Comment (peppe-hmrc) [reply app/uk/gov/hmrc/ctc/models/data/Transit.scala]: Personally I was agreeing with you about the `_id` but I saw it in place and I'm not very convinced. This syntax get very confusing with scala char `_` ... I don't remember where I saw it but there was something like `_._id` which is like the syntax to access to a tuple.
However I see the advantage on having a shorter converter ... if everybody agree I'm happy to change it

Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/models/data/Transit.scala]: We should talk about this more as a team and decide

File: test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala [hunk 5]
```diff
@@ -20,54 +20,117 @@ import org.scalatest.Matchers
...
+          await(transitRepository.getOldestBacklogTransit(transit.destinationPort, noTimeout))
+
+          val availableStaff = await(transitRepository.getAvailableStaff("backlog", noTimeout))
+
+          availableStaff shouldBe List.empty[AvailableStaffByPort]
+      }
+    }
   }
 
   "getPortCounts" should {
```
Comment (peppe-hmrc) [inline test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: Do you think we should add some tests to see and document that the priority is different depending if is `arrived` or `arriving`?
Have the feeling that there is so much logic inside the `aggregateData` which probably worth test and also it might simplify to understand what actually is and why?
This process of writing test is definitely not natural, but if you think when you wrote the code I'm confident you tested this logic, at least visually in mongo shell ... This is where I like TDD, reduce the arse of writing test later :)

File: test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala (hunk 1 above)
Comment (robertbutacu) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: After running the tests 5-10 times, it seems that I had no issues - since I override the timeout seconds to be 0, the only edge case would be when the createdDateTime and Instant.now are exactly the same to the millisecond. Having the Thread.sleep might've been too much. They are now commented, and should we have issues on the long run, we will have a (possible) reason for that.

Comment (peppe-hmrc) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: My point was not to remove the only the sleep, but insert an expired record to test the scenario. But looks like is too much work ... if removing the sleep will introduce faking test better having it :)

File: test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala (hunk 5 above)
Comment (robertbutacu) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: Yes, adding some more tests for different cases might be a good idea.

This should however happen on the Service level imo - there is some logic there which modifies the priority based on the business rules

Comment (peppe-hmrc) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: oh yes definitely sorry. I was referring to the Service

File: test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala (hunk 1 above)
Comment (robertbutacu) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: The repository collects the expired data based on the timeoutSeconds parameter that is passed in - from my point of view, whether its 0 or 600, it shouldnt matter and the function should correctly identify and ignore/add expired records - with this objective, it is doing what it's supposed to

File: it/uk/gov/hmrc/ctc/TransitControllerISpec.scala (hunk 2 above)
Comment (robertbutacu) [reply it/uk/gov/hmrc/ctc/TransitControllerISpec.scala]: Tbh, as long as all the known ports are there, then I believe there are cases to be made for both approaches - for now however, I would like to get MVP done and focus on details later on if there aren't any real performance issues.

File: test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala (hunk 5 above)
Comment (robertbutacu) [reply test/uk/gov/hmrc/ctc/repositories/TransitRepositorySpec.scala]: Will add this by standup tomorrow

File: app/uk/gov/hmrc/ctc/handlers/ErrorHandler.scala (hunk 3 above)
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/handlers/ErrorHandler.scala]: With the amount of work going around, this kind of small renaming changes I believe are beneficial - but fair enough, point taken.

Next time there will be a separate pr

File: it/uk/gov/hmrc/ctc/TransitControllerISpec.scala (hunk 2 above)
Comment (robertbutacu) [reply it/uk/gov/hmrc/ctc/TransitControllerISpec.scala]: Alright - I thought about this more, and yes, I think it would be better from an API perspective to not show any 0s and include those details in the respective object ( will be a part of the FE adjustment model )
So, for enRoute and arrived, they both will contain the interval and max-index - I believe this would be better since they are both customizable and it only makes sense that they could backtrack the time arrival for a transit - are there any downsides to this?
